**Golden Section Method**

Approximate the minimum value of the function : 
\begin{equation}
  f(x) = exp(-x) + x^2
\end{equation}


In [2]:
# Import Libraries and setting the Golden ratio value
import math
R  = ((5 ** 0.5) - 1) / 2
import numpy as np

In [4]:
# function definition
def fun(x): return math.exp(-x) + x**2

#method to compute the minima by golden section 
def Golden_section(lb,ub,tol):
    iteration = 0          # counter to count the number of iterations 
    x1 = ub - ((ub-lb)*R)  # setting the x1 point by  golden ratio value property
    x2 = lb + ((ub-lb)*R)  # setting the x2 point by  golden ratio value property
    f1 = fun(x1)           
    f2 = fun(x2)
    print "Starting with Upper and Lower Bound ",ub,lb
    while (abs(ub-lb) > tol): #loop to check for the tolerance
        iteration += 1       
        if f1 < f2:           #condition to check for the location of minima, to the left
            ub = x2
            print round(ub,5),round(lb,5)
            x2 = x1           # set the values accordingly
            f2 = f1
            x1 = ub - ((ub-lb)*R)
            f1 = fun(x1)
        else:
            lb = x1           #condition otherwise
            print round(ub,5),round(lb,5)
            x1 = x2
            f1 = f2
            x2 = lb + ((ub-lb)*R)
            f2 = fun(x2)

    print "\nConverged after %g iterations \nPoint is : "%iteration,round(.5*(ub+lb),6)

In [5]:
Golden_section(-100,100,.001)

Starting with Upper and Lower Bound  100 -100
100.0 -23.6068
52.7864 -23.6068
23.6068 -23.6068
23.6068 -5.57281
12.46118 -5.57281
5.57281 -5.57281
5.57281 -1.31556
2.94169 -1.31556
1.31556 -1.31556
1.31556 -0.31056
0.69444 -0.31056
0.69444 0.07331
0.45719 0.07331
0.45719 0.21994
0.45719 0.31056
0.40118 0.31056
0.36657 0.31056
0.36657 0.33195
0.36657 0.34518
0.3584 0.34518
0.35335 0.34518
0.35335 0.3483
0.35335 0.35023
0.35216 0.35023
0.35216 0.35096
0.35216 0.35142

Converged after 26 iterations 
Point is :  0.351787


Comments on Result:
- The tolerance for the convergence is set to be **0.001**
- The upper bound is **100** and the lower bound is **-100**
- Using the above values, the golden section implementation converged after **26** iterations.
- Therefore, the point of minima is **0.351787**